### PR TITLE
fix: PUBACK packet not compatible with RabbitMQ

### LINKED ISSUE #92: Mqtt 5 puback is different from spec
```
3.4.2.2.1 Property Length
The length of the Properties in the PUBACK packet Variable Header encoded as a Variable Byte Integer. If the Remaining Length is less than 4 there is no Property Length and the value of 0 is used.
```

Spec says that if remaining len is < 4, there's no property length. I think 'there's no property length' implies that it's not part of the byte stream (beacause it checks out with mosquitto)

```js
const mqtt = require('mqtt-packet');

let puback = {
  cmd: 'puback',
  messageId: 42,
}

const opts = { protocolVersion: 5 }; // default is 4. Usually, opts is a connect packet
let buf = mqtt.generate(puback, opts)
console.log(buf.toString('hex').match(/../g).map((x) => `0x${x}`).join(', '))

```
 Output of the above code is 
```
0x40, 0x03, 0x00, 0x2a, 0x00
```

Output of ack from mosquitto with no reason code and properties is this

```
[
    0x40,
    0x2,
    0x0,
    0x9,
]
```

Is the last `0x00` necessary?

### PR DESCRIPTION
**Pull request by alaendle**
MQTT5 spec is not so clear about the possible remaining length of PUBACK pakets.

We have: 

- `If there are no properties, this MUST be indicated by including a Property Length of zero [MQTT-2.2.2-1]`.
This clearly allows us to have packets like [64, 4, 0 42, 0, 0] (puback, length:4, pakedId: 0 42, reason: 0, properties: 0]
- `The Client or Server sending the PUBACK packet MUST use one of the PUBACK Reason Codes [MQTT-3.4.2-1]. The Reason Code and Property Length can be omitted if the Reason Code is 0x00 (Success) and there are no Properties. In this case the PUBACK has a Remaining Length of 2.`
This allows packets like [64, 2, 0, 42]
- The discussion starts if it is valid to only have the reason byte without the properties length (due to the unclear statement in 
`3.4.2.2.1 Property Length
The length of the Properties in the PUBACK packet Variable Header encoded as a Variable Byte Integer.
If the Remaining Length is less than 4 there is no Property Length and the value of 0 is used.
` ) 
So basically - ist it valid to have [64, 3, 0, 42, 0]? I would argue no, the remaining length should be 2 or >= 4. So this patch strives to only generate packets with this remaining length. However we are still able to parse packets with a remaining length of 3.

To state my belief - if it were intended to be able to omit both bytes individually, it wouldn't makle sense limit this to reason code 0x00 - why should [64, 3, 0, 42, 0] be allowed, while [64, 3, 0, 42, 135] is not?

Fixes #92

### CODE REVIEW

**Comment by robertsLando** [comment]
@alaendle remember to add a test suite that covers the issue

**Comment by alaendle** [comment]
Please note that this is a breaking change on the wire - however it shouldn't be with regards to the specification. It could be further optimized by striving for length === 2 (but this would also need an additional complexity by checking the reason code). So I basically adapted the existing test code - because I fear there is no backwards compatible way to implement this. Tested against the "old" version of RabbitMQ that treated length === 3 as a protocol error.

File: writeToStream.js [hunk 1]
```diff
@@ -410,6 +410,7 @@ function confirmation (packet, stream, opts) {
   stream.write(protocol.ACKS[type][qos][dup][0])
 
   // Length
+  if (length === 3) length += 1
```
Comment (robertsLando) [inline writeToStream.js:413]: add comment to explain

File: writeToStream.js [hunk 2]
```diff
@@ -423,6 +424,10 @@ function confirmation (packet, stream, opts) {
   // properies mqtt 5
   if (propertiesData !== null) {
     propertiesData.write()
+  } else {
```
Comment (robertsLando) [inline writeToStream.js:427]: same

**Review by robertsLando — CHANGES_REQUESTED** [review]
I would  also add mqtt specs links to specific section in the comments.

Also could you explain me how this could be a breaking change?  Let's say I have a broker using mqtt-packet with this patch version and an mqtt client (using mqttjs that uses this library too) with an older mqtt-packet implementation. The broker sends the puback packet, it is encoded following this new standard, will mqttjs client be able to decode it with the older version?

If nope does this only happens when using mqtt5 protocol?

**Review by robertsLando — CHANGES_REQUESTED** [review]
I would  also add mqtt specs links to specific section in the comments.

Also could you explain me how this could be a breaking change?  Let's say I have a broker using mqtt-packet with this patch version and an mqtt client (using mqttjs that uses this library too) with an older mqtt-packet implementation. The broker sends the puback packet, it is encoded following this new standard, will mqttjs client be able to decode it with the older version?

If nope does this only happens when using mqtt5 protocol?

**Comment by alaendle** [comment]
Please note I have added a description of the PR just now - https://github.com/mqttjs/mqtt-packet/pull/142#issue-1788020845 - to make sure we discuss the same things 😄

File: test.js [hunk 3]
```diff
@@ -1868,18 +1884,32 @@ testParseError('Invalid header flag bits, must be 0x0 for puback packet', Buffer
...
+  0, 2 // Message ID
+]), { protocolVersion: 5 })
+
 testParseGenerate('puback with reason and no MQTT 5 properties', {
   cmd: 'puback',
   retain: false,
   qos: 0,
   dup: false,
-  length: 3,
+  length: 4,
```
Comment (robertsLando) [inline test.js:1905]: Shouldn't we also keep tha old test with length 3 and no user properties bit?

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Comment by robertsLando** [comment]
@mcollina Thoughts on this?

File: test.js (hunk 3 above)
Comment (alaendle) [reply test.js:1905]: You are right - this test is already in place - however the diff is misleading; take a closer look!
https://github.com/mqttjs/mqtt-packet/pull/142/files#diff-13876b4beb64b9f156474dc78f9c923952a7ca210d4507b6b3135bbe244f8a60R737-R751

Comment (alaendle) [reply test.js:1905]: @robertsLando, btw - thank you for taking the time to guide me through this - as already mentioned I'm absolutely no JS expert.

Comment (robertsLando) [reply test.js:1905]: Thanks to you for the PR :) BTW I saw it now, thanks!

**Review by robertsLando — COMMENTED** [review]
LGTM

**Review by robertsLando — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm
